File: typescript/base-hmi-types/src/screens/defaults/HmiDefaultProfile.ts
import { HmiScreenItemBase } from "../base/HmiScreenItemBase.js";

export class HmiDefaultProfile {
  private readonly values = new Map<string, unknown>();

  constructor(readonly name: string) {}

  set<T>(objectType: string, propertyName: string, value: T | undefined): void {
    this.values.set(getKey(objectType, propertyName), value);
  }

  tryGet<T>(item: HmiScreenItemBase, propertyName: string): { found: true; value: T | undefined } | { found: false } {
    if (item.hmiObjectType?.trim()) {
      const key = getKey(item.hmiObjectType, propertyName);
      if (this.values.has(key)) {
        return { found: true, value: this.values.get(key) as T | undefined };
      }
    }

    for (const objectType of getObjectTypeKeys(item)) {
      if (item.hmiObjectType === objectType) {
        continue;
      }

      const key = getKey(objectType, propertyName);
      if (this.values.has(key)) {
        return { found: true, value: this.values.get(key) as T | undefined };
      }
    }

    return { found: false };
  }
}

function getKey(objectType: string, propertyName: string): string {
  return `${objectType}\0${propertyName}`;
}

function getObjectTypeKeys(item: HmiScreenItemBase): string[] {
  const keys: string[] = [];
  let prototype = Object.getPrototypeOf(item);
  while (prototype !== undefined && prototype !== null) {
    const constructorName = prototype.constructor?.name;
    if (constructorName !== undefined && constructorName !== "Object") {
      keys.push(constructorName);
    }
    prototype = Object.getPrototypeOf(prototype);
  }
  return keys;
}

File: src/BaseHmiTypes/Screens/Defaults/HmiDefaultProfile.cs
using BaseHmiTypes.Screens.Base;

namespace BaseHmiTypes.Screens.Defaults;

public sealed class HmiDefaultProfile
{
    private static readonly Dictionary<Type, string[]> ObjectTypeKeyCache = new();
    private static readonly object ObjectTypeKeyCacheLock = new();

    private readonly Dictionary<HmiDefaultKey, object?> _values = new();

    public HmiDefaultProfile(string name)
    {
        Name = name;
    }

    public string Name { get; }

    public void Set<T>(string objectType, string propertyName, T? value)
    {
        _values[new HmiDefaultKey(objectType, propertyName)] = value;
    }

    public void Set<TItem, TValue>(string propertyName, TValue? value)
        where TItem : HmiScreenItemBase
    {
        Set(typeof(TItem).Name, propertyName, value);
    }

    public bool TryGet<T>(HmiScreenItemBase item, string propertyName, out T? value)
    {
        value = default;

        object? rawValue = null;
        var found = false;

        if (!string.IsNullOrWhiteSpace(item.HmiObjectType)
            && _values.TryGetValue(new HmiDefaultKey(item.HmiObjectType, propertyName), out rawValue))
        {
            found = true;
        }

        var objectTypeKeys = GetCachedObjectTypeKeys(item.GetType());
        for (var i = 0; !found && i < objectTypeKeys.Length; i++)
        {
            var objectType = objectTypeKeys[i];
            if (StringComparer.Ordinal.Equals(item.HmiObjectType, objectType))
                continue;

            found = _values.TryGetValue(new HmiDefaultKey(objectType, propertyName), out rawValue);
        }

        if (!found)
            return false;

        if (rawValue == null)
            return true;

        if (rawValue is T typedValue)
        {
            value = typedValue;
            return true;
        }

        return false;
    }

    private static string[] GetCachedObjectTypeKeys(Type itemType)
    {
        lock (ObjectTypeKeyCacheLock)
        {
            if (ObjectTypeKeyCache.TryGetValue(itemType, out var cachedKeys))
                return cachedKeys;

            var keys = new List<string>();
            for (var type = itemType; type != null && typeof(HmiScreenItemBase).IsAssignableFrom(type); type = type.BaseType)
                keys.Add(type.Name);

            cachedKeys = keys.ToArray();
            ObjectTypeKeyCache[itemType] = cachedKeys;
            return cachedKeys;
        }
    }

    private readonly struct HmiDefaultKey : IEquatable<HmiDefaultKey>
    {
        public HmiDefaultKey(string objectType, string propertyName)
        {
            ObjectType = objectType;
            PropertyName = propertyName;
        }

        public string ObjectType { get; }

        public string PropertyName { get; }

        public bool Equals(HmiDefaultKey other)
        {
            return StringComparer.Ordinal.Equals(ObjectType, other.ObjectType)
                && StringComparer.Ordinal.Equals(PropertyName, other.PropertyName);
        }

        public override bool Equals(object? obj)
        {
            return obj is HmiDefaultKey other && Equals(other);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return (StringComparer.Ordinal.GetHashCode(ObjectType) * 397)
                    ^ StringComparer.Ordinal.GetHashCode(PropertyName);
            }
        }
    }
}
